testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/group/GroupTest.java

```diff
@@ -956,6 +956,61 @@ public void searchForGroupsShouldOnlyReturnMatchingElementsOrIntermediatePaths()
         }
     }
 
+    /**
+     * Groups search with query returns unwanted groups
+     * @link https://github.com/keycloak/keycloak/issues/20445
+     */
+    @Test
+    public void searchForGroupsShouldOnlyReturnMatchingElementsOrIntermediatePaths2() {
+
+        /*
+         * /Test/SubGroup
+         * /Test1/SubGroup2/SubGroup3
+         * /Dummy
+         */
+        String needle = "Test";
+        GroupRepresentation gTest = GroupBuilder.create().name("Test").build();
+        GroupRepresentation gTestSubGroup = GroupBuilder.create().name("SubGroup").build();
+        GroupRepresentation gTest1 = GroupBuilder.create().name("Test1").build();
+        GroupRepresentation gTest1SubGroup2 = GroupBuilder.create().name("SubGroup2").build();
+        GroupRepresentation gTest1SubGroup2SubGroup3 = GroupBuilder.create().name("SubGroup3").build();
+        GroupRepresentation gDummy = GroupBuilder.create().name("Dummy").build();
+
+        RealmResource realm = adminClient.realms().realm(AuthRealm.TEST);
+
+        createGroup(realm, gTest);
+        createGroup(realm, gTest1);
+        createGroup(realm, gDummy);
+        addSubGroup(realm, gTest, gTestSubGroup);
+        addSubGroup(realm, gTest1, gTest1SubGroup2);
+        addSubGroup(realm, gTest1SubGroup2, gTest1SubGroup2SubGroup3);
+
+        try {
+            List<GroupRepresentation> result = realm.groups().groups(needle, 0, 100);
+
+            assertEquals(2, result.size());
+            assertEquals("Test", result.get(0).getName());
+            assertEquals(1, result.get(0).getSubGroups().size());
+            assertEquals("SubGroup", result.get(0).getSubGroups().get(0).getName());
+            assertEquals("Test1", result.get(1).getName());
+            assertEquals(1, result.get(1).getSubGroups().size());
+            assertEquals("SubGroup2", result.get(1).getSubGroups().get(0).getName());
+            assertEquals("SubGroup3", result.get(1).getSubGroups().get(0).getSubGroups().get(0).getName());
```

Comment:
Per [_I think the result should only contain the "last" matching node (deepest matching group in the path)._](https://github.com/keycloak/keycloak/issues/20445#issuecomment-1559294560), I would have thought that the result of search for `Test` would be:

```
/Test
/Test1
```

The test indicates that
```
/Test
  /SubGroup1
/Test1
  /SubGroup2
    /SubGroup3
```
is returned instead.

I can see the value of returning direct children, and while I have preference against it, I have no strong opinion here. FTR, it is not consistent with the behaviour expected per [KEYCLOAK-18380](https://issues.redhat.com/browse/KEYCLOAK-18380).

However this line mean that even indirect children are returned? If so, this could have performance and UX impact (see comments in [KEYCLOAK-18380](https://issues.redhat.com/browse/KEYCLOAK-18380)).

Am I missing something?